homeassistant/helpers/device_registry.py

```diff
@@ -512,12 +542,31 @@ def async_get_or_create(
         serial_number: str | None | UndefinedType = UNDEFINED,
         suggested_area: str | None | UndefinedType = UNDEFINED,
         sw_version: str | None | UndefinedType = UNDEFINED,
+        translation_key: str | None = None,
+        translation_placeholders: Mapping[str, str] | None = None,
         via_device: tuple[str, str] | None | UndefinedType = UNDEFINED,
     ) -> DeviceEntry:
         """Get device. Create if it doesn't exist."""
         if configuration_url is not UNDEFINED:
             configuration_url = _validate_configuration_url(configuration_url)
 
+        config_entry = self.hass.config_entries.async_get_entry(config_entry_id)
+        if config_entry is None:
+            raise HomeAssistantError(
+                f"Can't link device to unknown config entry {config_entry_id}"
+            )
+
+        if translation_key:
+            full_translation_key = (
+                f"component.{config_entry.domain}.device.{translation_key}.name"
+            )
+            translated_name = config_entry.translations.get(
+                full_translation_key, translation_key
+            )
```

Comment:
This won't work if one config entry is creating devices for another config entry, but I can't think of a case where this happens.
Also, instead of storing a reference to the translations in the config entry, we could just call `translations.async_get_cached_translations`. I think it's reasonably performant, and the case when getting translations for a single integration can easily be optimized further if needed.

Reply:
awaiting translation.async_get_translations is not likely to suspend at this point as everything should already be loaded so it probably doesn't make much difference.  I wouldn't worry about it for now. I'll do functional testing once the CI passes to make sure there is no new bottleneck.

Reply:
I didn't find any obvious new bottlenecks.